### PR TITLE
[4234] Make rows resizable in table representation 

### LINKED ISSUE #4234: Make row resizable in table representation
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### PR DESCRIPTION
**Pull request by jerome-obeo**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts [hunk 1]
```diff
@@ -52,19 +52,26 @@ export interface GQLTable {
...
+  width: number | undefined;
+  isResizable: boolean;
+  hidden: boolean;
 }
 
 export interface GQLLine {
   id: string;
   targetObjectId: string;
   cells: GQLCell[];
+  height?: number;
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts]: prefer `height: number | null`

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts [hunk 2]
```diff
@@ -52,19 +52,26 @@ export interface GQLTable {
...
+  isResizable: boolean;
 }
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts:69]: Isn't `initialHeight` missing?

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.tsx [hunk 3]
```diff
@@ -55,6 +65,7 @@ export const TableContent = memo(({ editingContextId, representationId, table, r
         sx: {
           backgroundColor: 'transparent', // required to remove the default mui backgroundColor that is defined as !important
           cursor: 'pointer',
+          height: row.original.height,
         },
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/TableContent.tsx]: Why don't use `intialHeight` has fallback ?

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.types.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { GQLLine, TableProps } from '../TableContent.types';
+
+export interface ResizeRowHandlerProps extends TableProps {
+  row: GQLLine;
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.types.ts]: I don't think it's a good idea to extend `TableProps`
In the future, there's no reason why these two types shouldn't diverge.

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import { ResizeRowHandlerProps } from './ResizeRowHandler.types';
+
+const useStyles = makeStyles()((theme) => ({
+  rowMain: {
+    '&:hover': {
+      backgroundColor: theme.palette.action.hover,
+    },
+  },
+  cell: {
+    '&:hover': {
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx]: These two types are never used

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    width: '24px',
+    height: '4px',
+    bottom: 0,
+    left: '15px',
+    cursor: 'row-resize',
+  },
+}));
+
+interface DragState {
+  isDragging: boolean;
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx]: Move this interface in `ResizeRowHandler.types.ts`

File: packages/tables/frontend/sirius-components-tables/src/graphql/mutation/useTableMutation.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  const [mutationResizeColumn, mutationResizeColumnResult] = useMutation<GQLResizeColumnData, GQLResizeColumnVariables>(
+    resizeColumnMutation
+  );
+  useReporting(mutationResizeColumnResult, (data: GQLResizeColumnData) => data.resizeTableColumn);
+
+  const [mutationResizeRow, mutationResizeRowResult] = useMutation<GQLResizeRowData, GQLResizeRowVariables>(
+    resizeRowMutation
+  );
+  useReporting(mutationResizeRowResult, (data: GQLResizeRowData) => data.resizeTableRow);
+
```
Comment (frouene) [inline packages/tables/frontend/sirius-components-tables/src/graphql/mutation/useTableMutation.ts]: Move this part to be on the same location that `resizeRow`

**Review by frouene — DISMISSED** [review]

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts (hunk 1 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts]: Done

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts (hunk 2 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/TableContent.types.ts:69]: Done

File: packages/tables/frontend/sirius-components-tables/src/table/TableContent.tsx (hunk 3 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/TableContent.tsx]: Done

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.types.ts (hunk 4 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.types.ts]: Done

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx (hunk 5 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx]: Done

File: packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx (hunk 6 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/table/row/ResizeRowHandler.tsx]: Done

File: packages/tables/frontend/sirius-components-tables/src/graphql/mutation/useTableMutation.ts (hunk 7 above)
Comment (jerome-obeo) [reply packages/tables/frontend/sirius-components-tables/src/graphql/mutation/useTableMutation.ts]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/forms/FormWithTableEditingContextDescriptionProvider.java [hunk 8]
```diff
@@ -140,6 +140,8 @@ private TableWidgetDescription getTableWidgetDescription() {
                 .headerLabelProvider(variableManager -> "")
                 .headerIconURLsProvider(variableManager -> List.of())
                 .headerIndexLabelProvider(variableManager -> "")
+                .isResizablePredicate(variableManager -> false)
+                .initialHeightProvider(variableManager -> 0)
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/forms/FormWithTableEditingContextDescriptionProvider.java:145]: I would rather use -1 in the example to indicate that it should be ignored

File: packages/sirius-web/backend/sirius-web-papaya/src/main/java/org/eclipse/sirius/web/papaya/representations/table/ProjectTableRepresentationDescriptionProvider.java [hunk 9]
```diff
@@ -72,6 +72,8 @@ public List<IRepresentationDescription> getRepresentationDescriptions(IEditingCo
                 .headerLabelProvider(variableManager -> "")
                 .headerIconURLsProvider(variableManager -> List.of())
                 .headerIndexLabelProvider(variableManager -> "")
+                .isResizablePredicate(variableManager -> true)
+                .initialHeightProvider(variableManager -> 53)
```
Comment (sbegaudeau) [inline packages/sirius-web/backend/sirius-web-papaya/src/main/java/org/eclipse/sirius/web/papaya/representations/table/ProjectTableRepresentationDescriptionProvider.java]: This example tends not to use any fancy option, so I'll make it non resizable

**Comment by sbegaudeau** [comment]
I'll rebase and merge it

**Comment by sbegaudeau** [comment]
I'll perform the (final?) rebase, check it a little and merge it

**Review by sbegaudeau — APPROVED** [review]
I'll merge it after the build

**Pull request merged**